mitodl/mitxonline
# feat: add ability to retire users on edX

# What are the relevant tickets?




Fixes https://github.com/mitodl/mitxonline/issues/697

# Description (What does it do?)

This PR updates the retire_users management command to initiate user retirement on edX. Currently, the command retires a user in MITxOnline and we need to run an extra command in edX to initiate the retirement.

# How can this be tested?


- Checkout this branch in MITxOnline locally.
- Clone and Checkout [edx-api-client](https://github.com/mitodl/edx-api-client) branch `asad/bulk-retirement-api`
- Install the local clone of edx-api-client in MITxOnline by using `pip install -e path/to/edx-api-client`
- Start edX locally, go to the lms shell and run `./manage.py lms populate_retirement_states` to create retirement states.
- Now run the following in lms shell
  - `app_name=retirement`
  - `user_name=retirement_service_worker`
  - `./manage.py lms manage_user $user_name $user_name@example.com --staff --superuser`
  - `./manage.py lms --settings=<your-settings> create_dot_application $app_name $user_name`
  - Copy the client id and secret for later.
- Update setting in edX, which can be done in private.py or lms.yml `RETIREMENT_SERVICE_WORKER_USERNAME = 'retirement_service_worker'`
- Clone https://github.com/mitodl/mitxpro-openedx-extensions and update line 22 in `mitxpro_core/settings/common.py` to `r"^/(admin|auth|login|logout|register|api|oauth2|user_api)", r"^/v1/accounts/bulk_retire_users"`
  - now build a pip-installable package using the guidelines provided in the [readme](https://github.com/mitodl/mitxpro-openedx-extensions).
  - copy and paste the .tar.gz file into edx-platform and install the package in the LMS shell by using `pip install mitxpro-openedx-extensions-1.0.0.tar.gz`.
- Now update the MITxOnline .env file and add the Client ID and Secret copied from the LMS shell.
  - `OPENEDX_RETIREMENT_SERVICE_WORKER_CLIENT_ID=<client_id>`
  - `OPENEDX_RETIREMENT_SERVICE_WORKER_CLIENT_SECRET=<client_secret>`
- We are done with all of the configurations.
- Create a user that we want to retire.
- Go to the MITxOnline web container and run `./manage.py retire_users --user <username>`. Output should be similar to the SS
  <img width="1344" alt="Screen Shot 2023-08-08 at 1 05 33 PM" src="https://github.com/mitodl/mitxonline/assets/52656433/fcfa53b8-b676-4aad-bd66-15ad75ce1523">
- Go to the LMS Django Admin at `/admin/user_api/userretirementstatus/`. Verify that a retirement status exists with Pending state for the user.


## Checklist:
- [x] Publish new version of edx-api-client
- [x] Merge https://github.com/mitodl/salt-ops/pull/1574 and https://github.com/mitodl/ol-infrastructure/pull/1667
- [x] Create retirement service worker and OAUTH2 application in edX
- [x] Add OPENEDX_RETIREMENT_SERVICE_WORKER_CLIENT_ID and OPENEDX_RETIREMENT_SERVICE_WORKER_CLIENT_SECRET to the vault so that they can be used at https://github.com/mitodl/salt-ops/pull/1574/files#diff-70bbf14162dbce0218569098f3d487edbd7944ac86ad4fbeb9d5af8429364c1cR134-R135
